receiver/k8sclusterreceiver/collection/metadata.go

```diff
@@ -25,42 +25,178 @@ import (
+	ResourceIDKey string
+	// ResourceID is the Kubernetes UID of the resource. In case of
+	// containers, this value is the container id.
+	ResourceID ResourceID
+	MetadataDelta
+}
+
+// GetKubernetesMetadataUpdate processes kubernetes metadata updates and returns
+// a map of a delta of metadata mapped to each resource.
+func GetKubernetesMetadataUpdate(old, new map[ResourceID]*KubernetesMetadata) []*KubernetesMetadataUpdate {
+	var out []*KubernetesMetadataUpdate
+
+	for id, oldMetadata := range old {
+		// if an object with the same id has a previous revision, take a delta
+		// of the metadata.
+		if newMetadata, ok := new[id]; ok {
+			if metadataDelta := getMetadataDelta(oldMetadata.metadata, newMetadata.metadata); metadataDelta != nil {
+				out = append(out, &KubernetesMetadataUpdate{
+					ResourceIDKey: oldMetadata.resourceIDKey,
+					ResourceID:    id,
+					MetadataDelta: *metadataDelta,
+				})
+			}
+		}
+	}
+
+	// In case there are resources in the current revision, that was not in the
+	// previous revision, collect metadata to be added
+	for id, km := range new {
+		// if an id is seen for the first time, all metadata need to be added.
+		if _, ok := old[id]; !ok {
+			out = append(out, &KubernetesMetadataUpdate{
+				ResourceIDKey: km.resourceIDKey,
+				ResourceID:    id,
+				MetadataDelta: MetadataDelta{MetadataToAdd: km.metadata},
+			})
+		}
+	}
+
+	return out
+}
+
+// MetadataDelta keeps track of changes to metadata on resources.
+// The fields on this struct should help determine if there have
+// been changes to resource metadata such as Kubernetes labels.
+// An example of how this is used. Let's say we are dealing with a
+// Pod that has the following labels -
+// {"env": "test", "team": "otell", "usser": "bob"}. Now, let's say
+// there's an update to one or more labels on the same Pod and the
+// labels now look like the following -
+// {"env": "test", "team": "otel", "user": "bob"}. The k8sclusterreceiver
+// upon receiving the event corresponding to the labels updates will
+// generate a MetadataDelta with the following values -
+// 					MetadataToAdd: {"user": "bob"}
+// 					MetadataToRemove: {"usser": "bob"}
+// 					MetadataToUpdate: {"team": "otel"}
+// Apart from Kubernetes labels, the other metadata collected by this
+// receiver are also handled in the same manner.
+type MetadataDelta struct {
+	// MetadataToAdd contains key-value pairs that are newly added to
+	// the resource description in the current revision.
+	MetadataToAdd map[string]string
+	// MetadataToRemove contains key-value pairs that no longer describe
+	// a resource and needs to be removed.
+	MetadataToRemove map[string]string
+	// MetadataToUpdate contains key-value pairs that have been updated
+	// in the current revision compared to the previous revisions(s).
+	MetadataToUpdate map[string]string
+}
+
+// getMetadataDelta returns MetadataDelta between two sets for properties
```

Comment:
Would be nice to tell here that nil is returned when old and new are equal and delta is empty.

Reply:
Done.